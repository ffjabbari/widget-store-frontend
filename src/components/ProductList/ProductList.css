@media (min-width: 576px) {
  .card-deck {
    margin-right: 0px;
    margin-left: 0px;
  }
  .card-deck .card {
    margin-right: 5px;
    margin-left: 5px;
  }
}

@media (min-width: 1200px) {
  .card-deck {
    margin-right: -15px;
    margin-left: -15px;
  }
  .card-deck .card {
    margin-right: 15px;
    margin-left: 15px;
  }
}
