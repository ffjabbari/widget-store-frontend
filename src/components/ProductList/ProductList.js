import React from 'react';
import { CardDeck } from 'reactstrap';

// local imports
import APIRequest from '../../hooks/apiHook';
import ProductCard from '../ProductCard/ProductCard';

function ProductList({ params }) {
  const { loading, error, response } = APIRequest(
    '/products',
    'get',
    null,
    params
  );

  if (error) return <div>Error</div>;
  if (loading) return <div>loading...</div>;

  return (
    <CardDeck className="justify-content-center pb-2">
      {response.products.map(product => (
        <ProductCard product={product} key={product.product_id} />
      ))}
    </CardDeck>
  );
}

export default React.memo(ProductList);
