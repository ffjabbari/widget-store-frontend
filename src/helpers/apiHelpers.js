/** API Helper functions. */

// local imports
import { asyncAPIRequest } from '../hooks/apiHook';

/**
 * updateUserProfile()
 *
 * Make patch request to update a user profile. Used by CustomerAddressForm
 * to allow updating customer address of signed in users.
 *
 * @param {object} values - form values
 * @param {object} user - user object
 * @param {token} token - user token
 * @param {function} setResponseError - state setter for error display
 * @param {any} defaultAddress - Falsey or obj with address components
 * @param {function} setFormDisabled - state setter to hide form after update
 *
 * returns: true if an error happened else undefined
 */

export const updateUserProfile = async (
  values,
  user,
  token,
  setResponseError,
  defaultAddress,
  setFormDisabled
) => {
  const userUpdateData = { ...values };
  // remove email to dis-allow email updating from here
  delete userUpdateData['email'];
  delete userUpdateData['user_id'];

  // remomve key/value pairs having a value of empty string from userUpdateData
  Object.entries(userUpdateData).forEach(([key, value]) => {
    if (value === '') delete userUpdateData[key];
  });

  try {
    // make patch request
    const resp = await asyncAPIRequest(
      `/users/${user.username}`,
      'patch',
      userUpdateData,
      { _token: token }
    );
    if (resp.error) {
      setResponseError(resp.error.response.data.message);
      return true;
    } else {
      defaultAddress.current = values;
      setFormDisabled(true);
    }
  } catch (error) {
    setResponseError(error.message);
    return true;
  }
};
