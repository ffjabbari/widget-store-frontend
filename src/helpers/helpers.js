import { Redirect } from 'react-router-dom';
import { motion } from 'framer-motion';

/**
 * Given a path return the "root" argument.
 * For example the "root" of /shop/product/206 would be "/shop".
 *
 * @param {string} path
 */

export const getPathRoot = path => {
  return path.match(/^(\/\w*)\/?.*$/)[1];
};

export const FramerRedirect = props => (
  <motion.div exit="undefined">
    <Redirect {...props} />
  </motion.div>
);

export const animateVariant = (setterFunct, delay, callback, args = []) => {
  // if animation state is false set to 'default' else 'active'
  setterFunct(state => (state === false ? 'default' : 'active'));
  setTimeout(() => {
    setterFunct('default');
    if (callback) callback(...args);
  }, delay);
};
