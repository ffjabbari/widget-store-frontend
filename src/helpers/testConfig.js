/** Test data setup/teardown helper hooks */

export const TEST_DATA = {};

export const populateTestDataHook = testData => {
  testData.product = {
    byline: 'Latest generation smart televison 2021',
    date_added: '2020-12-29T08:00:00.000Z',
    description:
      'Lorem ipsum dolor sit, amet consectetur adipisicing elit. Nihil, voluptates enim nam voluptatum, quia doloribus quis sequi, porro consequatur excepturi eos error voluptas nobis modi atque magni commodi amet placeat.',
    discount: '0.00',
    image_url:
      'https://www.lg.com/sg/images/tvs/md05271244/gallery/LG-43LH600T-L.jpg',
    name: 'Samsung XG-900 55 inch LCD TV',
    price: '400.40',
    product_id: 1,
    quantity: 100,
    rating: null,
  };

  testData.product2 = {
    byline: 'A good remote.',
    discount: '0.15',
    image_url:
      'https://www.lg.com/sg/images/tvs/md05271244/gallery/LG-43LH600T-L.jpg',
    name: 'Logitech Singularity Universal Remote',
    price: '40.20',
    product_id: 2,
    rating: null,
    quantity: 1,
  };

  testData.userProfileData = {
    data: {
      user: {
        user_id: 1,
        username: 'widgetLord',
        email: 'foo@gmail.com',
        first_name: 'bob',
        last_name: 'bobert',
        address: '123 Hill Ave.',
        address_line2: null,
        city: 'Big City',
        state: 'Texas',
        postal_code: '33333',
        phone_number: null,
        avatar_url: null,
        orders: [
          {
            order_id: 1,
            order_date: '2021-01-24T08:00:00.000Z',
            status: 'Pending',
          },
        ],
      },
    },
  };

  testData.CustomerInfoFormsValues = {
    first_name: 'Test',
    last_name: 'User',
    email: 'foo@gmail.com',
    address: '123 Main St',
    address_line2: '',
    city: 'Big City',
    state: 'TX',
    postal_code: '39488',
    phone_number: '(555) 555-5555',
    set_default: false,
  };

  testData.cart = {
    items: {
      2: {
        byline: 'The only remote you will need',
        date_added: '2021-01-24T08:00:00.000Z',
        description: 'Lorem ipsum dolor sit, placeat.',
        discount: '0.15',
        image_url:
          'https://cnet3.cbsistatic.com/img/SB-Lc9lni02k-GQeFZlVGII6cpg=/1070x602/2010/08/03/57369690-bb77-11e2-8a8e-0291187978f3/33981226-2-1333-OVR-1.jpg',
        name: 'Logitech Singularity Universal Remote',
        price: '40.20',
        product_id: 2,
        quantity: 2,
        rating: null,
      },
    },
    numCartItems: 1,
    subtotal: '34.17',
  };
};
