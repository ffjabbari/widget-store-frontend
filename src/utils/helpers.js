import { Redirect, Link } from 'react-router-dom';
import { motion } from 'framer-motion';

/**
 * getPathRoot()
 *
 * Given a path return the "root" argument.
 * For example the "root" of /shop/product/206 would be "/shop".
 *
 * @param {string} path
 */

export const getPathRoot = path => {
  return path.match(/^(\/\w*)\/?.*$/)[1];
};

/**
 * <FramerRedirect to="" />
 *
 * Component to force react-router-dom redirect when framer-motion
 * prevents a redirect from executing.
 *
 * If regular <Redirect /> fails in this environment use this redirect instead.
 *
 * @param {any} props
 *
 * returns: motion.div wrapped redirect component
 */
export const FramerRedirect = props => (
  <motion.div exit="undefined">
    <Redirect {...props} />
  </motion.div>
);
export const FramerLink = props => (
  <motion.div exit="undefined">
    <Link {...props} />
  </motion.div>
);

/**
 * animateVariant()
 *
 * Animation function to turn on an animation state then turn it back
 * off after a delay. The state will be set "active" then back to "default".
 *
 * If inital state provided by setterFunct is false this function will skip
 * setting the animation state "active" instead set it to "default".
 *
 * If a callback is provided it will be called with any arguments
 * after setting state back to "default".
 *
 *
 * @param {function} setterFunct - useState setter funct to set animation state
 * @param {integer} delay - delay of setterFunct being called
 * @param {funct} callback - function to call upon completion
 * @param {array} args - array of arguments for callback function
 */

export const animateVariant = (setterFunct, delay, callback, args = []) => {
  // if animation state is false set to 'default' else 'active'
  setterFunct(state => (state === false ? 'default' : 'active'));
  // after delay set back to "default"
  return setTimeout(() => {
    setterFunct('default');
    if (callback) callback(...args);
  }, delay);
};
