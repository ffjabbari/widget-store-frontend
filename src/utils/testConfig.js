/** Test data setup/teardown helper hooks */

export const TEST_DATA = {};

export const populateTestDataHook = testData => {
  testData.product = {
    byline: 'Latest generation smart televison 2021',
    date_added: '2020-12-29T08:00:00.000Z',
    description:
      'Lorem ipsum dolor sit, amet consectetur adipisicing elit. Nihil, voluptates enim nam voluptatum, quia doloribus quis sequi, porro consequatur excepturi eos error voluptas nobis modi atque magni commodi amet placeat.',
    discount: '0.00',
    image_url:
      'https://www.lg.com/sg/images/tvs/md05271244/gallery/LG-43LH600T-L.jpg',
    name: 'Samsung XG-900 55 inch LCD TV',
    price: '400.40',
    product_id: 1,
    quantity: 100,
    rating: null,
  };

  testData.product2 = {
    byline: 'The only remote you will need',
    description: 'Lorem ipsum dolor sit, placeat.',
    discount: '0.15',
    image_url:
      'https://cnet3.cbsistatic.com/img/SB-Lc9lni02k-GQeFZlVGII6cpg=/1070x602/2010/08/03/57369690-bb77-11e2-8a8e-0291187978f3/33981226-2-1333-OVR-1.jpg',
    name: 'Logitech Singularity Universal Remote',
    price: '40.20',
    product_id: 2,
    quantity: 1,
    rating: null,
  };

  testData.userProfileData = {
    data: {
      user: {
        user_id: 1,
        username: 'widgetLord',
        email: 'foo@gmail.com',
        first_name: 'bob',
        last_name: 'bobert',
        address: '123 Hill Ave.',
        address_line2: null,
        city: 'Big City',
        state: 'Texas',
        postal_code: '33333',
        phone_number: null,
        avatar_url: null,
        orders: [
          {
            order_id: 1,
            order_date: '2021-01-24T08:00:00.000Z',
            status: 'Pending',
          },
        ],
      },
    },
  };

  testData.CustomerInfoFormsValues = {
    first_name: 'Test',
    last_name: 'User',
    email: 'foo@gmail.com',
    address: '123 Main St',
    address_line2: 'Apt 5',
    city: 'Big City',
    state: 'TX',
    postal_code: '39488',
    phone_number: '(555) 555-5555',
  };

  testData.cart = {
    items: {
      2: {
        byline: 'The only remote you will need',
        description: 'Lorem ipsum dolor sit, placeat.',
        discount: '0.15',
        image_url:
          'https://cnet3.cbsistatic.com/img/SB-Lc9lni02k-GQeFZlVGII6cpg=/1070x602/2010/08/03/57369690-bb77-11e2-8a8e-0291187978f3/33981226-2-1333-OVR-1.jpg',
        name: 'Logitech Singularity Universal Remote',
        price: '40.20',
        product_id: 2,
        quantity: 1,
        rating: null,
      },
    },
    numCartItems: 1,
    subtotal: '34.17',
  };

  // mock return from /order/:id
  testData.orderGet = {
    customer: 1,
    customer_info: {
      first_name: 'Test',
      last_name: 'User',
      email: 'foo@gmail.com',
      address: '123 Main St',
      address_line2: '',
      city: 'Big City',
      state: 'TX',
      postal_code: '39488',
      phone_number: '(555) 555-5555',
    },

    discount: null,
    items: [
      { 0: { order_id: 2, product_id: 6, quantity: 1 } },
      { 1: { order_id: 2, product_id: 8, quantity: 1 } },
    ],

    order_date: '2021-02-05T08:00:00.000Z',
    order_id: 2,
    shipping_address: null,
    shipping_cost: '12.00',
    shipping_method: {
      details: { cost: '12.00', name: 'USPS ground shipping (3-7 days)' },
      shipping_method: 'usps_ground',
    },

    status: 'Pending',
    subtotal: '784.16',
    tax: '66.65',
    total: '862.81',
    total_items_quantity: 2,
    tracking_number: null,
  };

  TEST_DATA.shippingMethods = {
    usps_ground: {
      cost: '12.00',
      name: 'USPS ground shipping (3-7 days)',
    },
    usps_priority: {
      cost: '16.00',
      name: 'USPS priority shipping (2-4 days)',
    },
    usps_next_day: {
      cost: '25.00',
      name: 'USPS next day delivery (1 day)',
    },
    ups_ground: {
      cost: '16.00',
      name: 'UPS ground shipping (3-6 days)',
    },
    ups_2nd_day: {
      cost: '20.00',
      name: 'UPS second day delivery (2 days)',
    },
    ups_overnight: {
      cost: '28.00',
      name: 'UPS overnight delivery (1 day)',
    },
  };
};
