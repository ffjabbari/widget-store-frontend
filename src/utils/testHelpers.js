import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { testStore } from '../redux/store/reduxStore';

export const renderWithStore = component =>
  render(
    <Provider store={testStore}>
      <MemoryRouter>{component}</MemoryRouter>
    </Provider>
  );
