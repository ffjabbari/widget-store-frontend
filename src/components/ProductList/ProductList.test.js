/** ProductList tests */

// local imports
import { act } from '@testing-library/react';
import { renderWithStore } from '../../helpers/testHelpers';
import { TEST_DATA, populateTestDataHook } from '../../helpers/testConfig';
import ProductList from './ProductList';

const axios = require('axios');
jest.mock('axios');

beforeAll(() => {
  populateTestDataHook(TEST_DATA);
});

beforeEach(() => {
  axios.mockResolvedValue({
    data: {
      products: [TEST_DATA.product],
    },
  });
});

afterEach(() => {
  jest.resetAllMocks();
});

test('renders ProductList', async () => {
  await act(async () => {
    renderWithStore(<ProductList />);
  });
});

test('ProductList snapshot', async () => {
  let asFragment;
  await act(async () => {
    const resp = renderWithStore(<ProductList />);
    asFragment = resp.asFragment;
  });
  expect(asFragment()).toMatchSnapshot();
});

test('Product name in document', async () => {
  let getByText;
  await act(async () => {
    const resp = renderWithStore(<ProductList />);
    getByText = resp.getByText;
  });
  const cardText = getByText('Samsung XG-900 55 inch LCD TV');
  expect(cardText).toBeInTheDocument();
});
