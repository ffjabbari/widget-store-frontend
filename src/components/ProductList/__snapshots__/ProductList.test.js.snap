// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList snapshot 1`] = `
<DocumentFragment>
  <div
    class="justify-content-center pb-2 card-deck"
  >
    <div
      class="ProductCard mb-4 mx-auto card"
    >
      <div
        class="img-wrapper"
      >
        <img
          alt="Samsung XG-900 55 inch LCD TV"
          class="card-img-top"
          height="100%"
          src="https://www.lg.com/sg/images/tvs/md05271244/gallery/LG-43LH600T-L.jpg"
          width="auto"
        />
      </div>
      <div
        class="card-body"
      >
        <h5
          class="card-title"
        >
          Samsung XG-900 55 inch LCD TV
        </h5>
        <h6
          class="mb-2 text-muted card-subtitle"
        >
          <em>
            Latest generation smart televison 2021
          </em>
        </h6>
        <p
          class="card-text"
        >
          <span
            class="sc-bdfBwQ ehKgzr"
          >
            <span
              class="sc-gsTCUz cxQGHo price"
            >
              $400.40
            </span>
          </span>
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
